.github/workflows/project_filter.yml

```diff
@@ -0,0 +1,92 @@
+        include:
+          - appid: nf01
+            apppath: 01-indoor-floor-level-tracker
+            pio_project: firmware
+            shell_script: ''
+          - appid: nf04
+            apppath: 04-low-power-digital-signage
+            pio_project: firmware
+          - appid: nf08
+            apppath: 08-power-quality-monitor
+            pio_project: firmware
+          - appid: nf09
+            apppath: 09-valve-monitor
+            pio_project: firmware/arduino
+          - appid: nf10
+            apppath: 10-flow-rate-monitor
+            pio_project: firmware
+          - appid: nf11
+            apppath: 11-generator-activity-monitor
+            pio_project: firmware
+          - appid: nf12
+            apppath: 12-remote-power-control
+            pio_project: firmware
+          - appid: nf13
+            apppath: 13-tool-usage-cycle-tracking
+            pio_project: firmware
+          - appid: nf15
+            apppath: 15-greenhouse-monitor
+            pio_project: firmware
+          - appid: nf18
+            apppath: 18-temperature-and-humidity-monitor
+            pio_project: firmware/arduino
+          - appid: nf30
+            apppath: 38-audio-classifier
+            pio_project: classifier
+          - appid: nf30
+            apppath: 38-audio-classifier
+            pio_project: data_acquisition
+          - appid: nf34
+            apppath: 34-cellular-modbus-client
+            pio_project: firmware
+          - appid: nf35
+            apppath: 35-CAN-vehicle-monitor
+            pio_project: firmware
+          - appid: nf36
+            apppath: 36-vending-machine-monitor
+            pio_project: firmware
+          - appid: nf38
+            apppath: 38-audio-classifier
+          - appid: nf41
+            apppath: 41-cold-chain-monitor
+            pio_project: firmware/arduino
+          - appid: nf43
+            apppath: 43-propane-tank-fuel-gauge
+            pio_project: firmware
+          - appid: sense-hat-reference
+            apppath: sense-hat-reference
+            pio_project: .
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          submodules: false
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          # uncomment for local testing
+          #          base: HEAD
+          filters: |
+            app:
+              - ${{ matrix.apppath }}/**
+      - name: "Build Platform IO project"
+        if: steps.filter.outputs.app == 'true' && matrix.pio_project != ''
+        uses: "./.github/actions/platformio"
+        with:
+            project_dir: "${{ matrix.apppath }}/${{ matrix.pio_project }}"
+
+      - name: "Run App CI shell script"
+        if: steps.filter.outputs.app == 'true' && matrix.shell_script != ''
+        # uses must be a literal. can't do dynamic lookup of the action. So we use shell scripts instead.
+        run: "${{ matrix.shell_script }} ${{ matrix.apppath }}"
```

Comment:
I don't quite understand what's happening here. What's `shell_script`? I only see it defined for nf01, and it's an empty string.

Reply:
I originally chose to use shell scripts because you cannot dynamically dispatch to an action (the name of the file cannot be interpolated, and what I ideally wanted was to run a `ci/action.yml` file from within `apppath`. 

Fortunately, for the platformIO projects, they all use the same action, so there wasn't much repetition involved, and switching to actions worked fine in that case.  I left in the shell script support  in case it's useful.

Reply:
I ideally wanted to add `shell_script` at the global level, default value empty, but that caused problems. Defining it in the matrix seems to avoid that.  Not ideal, but didn't want to spend time digging deeper.